File: android-java/app/src/main/java/com/oursky/skeleton/ui/base/BaseFrameLayout.java
package com.oursky.skeleton.ui.base;

import android.content.Context;
import android.support.annotation.NonNull;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;

@SuppressWarnings({"WeakerAccess", "unused"})
public abstract class BaseFrameLayout extends FrameLayout {
    // delegate
    protected abstract void onAttach(@NonNull View view);
    protected abstract void onDetach(@NonNull View view);

    //region Lifecycle
    //---------------------------------------------------------------
    public BaseFrameLayout(Context context) {
        super(context);
    }
    public BaseFrameLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }
    public BaseFrameLayout(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }
    //---------------------------------------------------------------
    //endregion

    //region Attach / Detach
    //---------------------------------------------------------------
    @Override
    final public void onAttachedToWindow() {
        super.onAttachedToWindow();
        onAttach(this);
    }
    @Override
    final public void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        onDetach(this);
    }
    //---------------------------------------------------------------
    //endregion
}

File: android-kotlin/app/src/main/java/com/oursky/skeleton/ui/base/BaseFrameLayout.kt
package com.oursky.skeleton.ui.base

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout

abstract class BaseFrameLayout : FrameLayout {
    // delegate
    protected abstract fun onAttach(view: View)
    protected abstract fun onDetach(view: View)

    //region Lifecycle
    //---------------------------------------------------------------
    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)
    //---------------------------------------------------------------
    //endregion

    //region Attach / Detach
    //---------------------------------------------------------------
    final override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        onAttach(this)
    }
    final override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        onDetach(this)
    }
    //---------------------------------------------------------------
    //endregion
}
